### PR TITLE
Add trace_id to log method

### PR DESCRIPTION
**Pull request by jamie256**
* add trace_id as optional parameter to log method
* generate a trace_id from uuid4 if not specified
* add metadata property to profiles and result sets
* add trace_id to the metadata
* set trace_id when uploading results
* Add some unit tests to cover new functionality
* Add result sets unit test.

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/tests/api/logger/test_segments.py [hunk 1]
```diff
@@ -94,6 +94,43 @@ def test_single_column_segment() -> None:
...
+    assert first_segment_profile is not None
+    assert first_segment_profile._columns["col1"]._schema.dtype == np.int64
+    assert first_segment_profile._columns["col2"]._schema.dtype == np.float64
+    assert first_segment_profile._columns["col3"]._schema.dtype.name == "object"
+    segment_cardinality: CardinalityMetric = (
+        first_segment_profile.view().get_column(segment_column).get_metric("cardinality")
+    )
+    cardinality = segment_cardinality.estimate
+    assert cardinality is not None
+    assert cardinality == 1.0
```
Comment (richard-rogers) [inline python/tests/api/logger/test_segments.py:131]: Took me a second to figure out why this was 1.0. Perhaps a comment? Could just be the jet lag :)

File: python/whylogs/api/logger/logger.py [hunk 2]
```diff
@@ -98,13 +102,22 @@ def log(
 
         # If segments are defined use segment_processing to return a SegmentedResultSet
         if active_schema and active_schema.segments:
-            return segment_processing(active_schema, obj, pandas, row, self._segment_cache)
+            segmented_results = segment_processing(active_schema, obj, pandas, row, self._segment_cache)
+            _extract_common_profile_metadata(
```
Comment (richard-rogers) [inline python/whylogs/api/logger/logger.py]: I think I'd be happier if the return value was assigned to something.

File: python/whylogs/api/logger/logger.py [hunk 3]
```diff
@@ -98,13 +102,22 @@ def log(
...
+                segmented_results.metadata, trace_id=trace_id, tags=tags, segment_key_values=segment_key_values
+            )
+            return segmented_results
 
         profiles = self._get_matching_profiles(obj, pandas=pandas, row=row, schema=active_schema)
 
         for prof in profiles:
             prof.track(obj, pandas=pandas, row=row, execute_udfs=False)
-            prof._metadata
```
Comment (richard-rogers) [inline python/whylogs/api/logger/logger.py:107]: Was old line 107 a bug?

File: python/whylogs/api/logger/segment_processing.py [hunk 4]
```diff
@@ -126,6 +126,7 @@ def segment_processing(
     pandas: Optional[pd.DataFrame] = None,
     row: Optional[Dict[str, Any]] = None,
     segment_cache: Optional[SegmentCache] = None,
+    trace_id: Optional[str] = None,
```
Comment (richard-rogers) [inline python/whylogs/api/logger/segment_processing.py]: Should we do something with this argument?

File: python/whylogs/core/metadata.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+diagnostic_logger = logging.getLogger(__name__)
+
+WHYLABS_TRACE_ID_KEY = "whylabs.traceId"
+CREATION_TIMESTAMP_KEY = "whylogs.creationTimestamp"
+DATASET_TIMESTAMP_KEY = "whylogs.datasetTimestamp"
+USER_TAGS_KEY = "whylogs.user.tags"
+NAME_KEY = "whylogs.name"
+
+
+def _extract_common_profile_metadata(
```
Comment (richard-rogers) [inline python/whylogs/core/metadata.py]: `s/extract/populate/`

File: python/whylogs/api/logger/segment_processing.py (hunk 4 above)
Comment (jamie256) [reply python/whylogs/api/logger/segment_processing.py]: Good catch, removing.

File: python/whylogs/core/metadata.py (hunk 5 above)
Comment (jamie256) [reply python/whylogs/core/metadata.py]: sure.

File: python/whylogs/api/logger/logger.py (hunk 3 above)
Comment (jamie256) [reply python/whylogs/api/logger/logger.py:107]: yes! A casualty of splitting previous PRs from larger changes with the _metadata.

File: python/whylogs/api/logger/logger.py (hunk 2 above)
Comment (jamie256) [reply python/whylogs/api/logger/logger.py]: This updates the dictionary. I added a comment in the code.

result_set has a metadata property but not a metadata setter. The idea is that we are aggregating information but not creating new instances of the collection on every modification.

File: python/tests/api/logger/test_segments.py (hunk 1 above)
Comment (jamie256) [reply python/tests/api/logger/test_segments.py:131]: added a comment.

**Review by richard-rogers — APPROVED** [review]

File: python/whylogs/api/logger/logger.py (hunk 2 above)
Comment (richard-rogers) [reply python/whylogs/api/logger/logger.py]: I might still do `_ = _populate_common_profile_metadate(...)  # I like side-effects :)`

**Pull request merged**